plugins/inputs/modbus/type_conversions.go

```diff
@@ -2,53 +2,27 @@ package modbus
 
 import "fmt"
 
-func determineConverter(inType, byteOrder, outType string, scale float64) (fieldConverterFunc, error) {
-	if scale != 0.0 {
-		return determineConverterScale(inType, byteOrder, outType, scale)
+func determineConverter(inType, byteOrder, outType string, scale float64, shift float64) (fieldConverterFunc, error) {
+	if scale == 0.0 {
+		scale = 1.0
 	}
-	return determineConverterNoScale(inType, byteOrder, outType)
-}
-
-func determineConverterScale(inType, byteOrder, outType string, scale float64) (fieldConverterFunc, error) {
-	switch inType {
-	case "INT16":
-		return determineConverterI16Scale(outType, byteOrder, scale)
-	case "UINT16":
-		return determineConverterU16Scale(outType, byteOrder, scale)
-	case "INT32":
-		return determineConverterI32Scale(outType, byteOrder, scale)
-	case "UINT32":
-		return determineConverterU32Scale(outType, byteOrder, scale)
-	case "INT64":
-		return determineConverterI64Scale(outType, byteOrder, scale)
-	case "UINT64":
-		return determineConverterU64Scale(outType, byteOrder, scale)
-	case "FLOAT32":
-		return determineConverterF32Scale(outType, byteOrder, scale)
-	case "FLOAT64":
-		return determineConverterF64Scale(outType, byteOrder, scale)
-	}
-	return nil, fmt.Errorf("invalid input data-type: %s", inType)
-}
-
-func determineConverterNoScale(inType, byteOrder, outType string) (fieldConverterFunc, error) {
```

Comment:
No. Your approach will **always** convert the input to a floating point number and, even worse, performs a multiplication before converting to the output type. This will lead to a loss of precision!!! Try
```golang
package main

import (
	"fmt"
)

func main() {
	a := ^uint64(0)
	fmt.Println(a)

	b := uint64(float64(a))
	fmt.Println(b)

	fmt.Println(a == b)
}
```

Reply:
This is **only** true for the (legacy) register-based configuration but **not** for the request-based configuration. The point here is that this _exactly_ reproduced previous behavior for register-based but improves stuff for request-based!